### PR TITLE
Fixes for the LLaMA backbone + add dropout

### PR DESCRIPTION
**Pull request by tirthasheshpatel**
Get the LLaMA backbone and its components closer to Mistral. Later, it would be nice to unify them to reduce code duplication. Tests haven't been modified, so the behavior of the components remain the same.

Note that I don't update the checkpoint file in this PR; it would be better to do it once the other components (preprocessor and generator) are in.

### CODE REVIEW

File: keras_nlp/models/llama/llama_attention.py [hunk 1]
```diff
@@ -17,35 +17,34 @@
...
         kernel_initializer="glorot_uniform",
-        rope_max_wavelength=10000,
-        max_sequence_length=512,
+        dropout=0,
         **kwargs,
     ):
         super().__init__(**kwargs)
-        self.num_query_heads = num_query_heads
-        self.num_key_value_heads = num_key_value_heads
+        self._num_query_heads = num_query_heads
```
Comment (mattdangerw) [inline keras_nlp/models/llama/llama_attention.py:34]: Can we keep the attrs that correspond to config/init args public? That is the style we want actually, even though we are not totally consistent here or in core Keras.

File: keras_nlp/models/llama/llama_attention.py [hunk 2]
```diff
@@ -86,21 +97,30 @@ def build(self, inputs_shape):
...
+        self._dropout_layer = keras.layers.Dropout(
+            rate=self._dropout,
+            dtype=self.dtype_policy,
+        )
+
         self._output_dense = keras.layers.EinsumDense(
-            equation="bqm,mh->bqh",
-            output_shape=(None, self.hidden_dim),
-            kernel_initializer=clone_initializer(self.kernel_initializer),
+            equation="bquh,uhm->bqm",
```
Comment (mattdangerw) [inline keras_nlp/models/llama/llama_attention.py:106]: How did this work before? Or did it not?

File: keras_nlp/models/llama/llama_decoder.py [hunk 3]
```diff
@@ -19,25 +19,25 @@
...
 
     def __init__(
         self,
         intermediate_dim,
         num_query_heads,
         num_key_value_heads,
+        rope_max_wavelength=10000,
         rope_scaling_factor=1.0,
-        activation="relu",
+        activation="silu",
```
Comment (mattdangerw) [inline keras_nlp/models/llama/llama_decoder.py:37]: Was this also just broken?

File: keras_nlp/models/llama/llama_attention.py [hunk 4]
```diff
@@ -17,35 +17,34 @@
 from keras_nlp.utils.keras_utils import clone_initializer
 
 
-class LlamaAttention(keras.layers.Layer):
-    """Grouped query attention for Llama models"""
+class CachedLlamaAttention(keras.layers.Layer):
```
Comment (mattdangerw) [inline keras_nlp/models/llama/llama_attention.py:20]: I don't actually mind `LlamaAttention` as a name here. I'd rather clean up the gemma side someday.

**Review by mattdangerw — COMMENTED** [review]
Looks good! Couple comments.

File: keras_nlp/models/llama/llama_attention.py (hunk 2 above)
Comment (tirthasheshpatel) [reply keras_nlp/models/llama/llama_attention.py:106]: It did work since we reshaped the inputs from (B, Q, U, H) -> (B, Q, U * H) before passing them to the output dense layer. Both the ways are valid. I just used the same thing I wrote for Mistral.

File: keras_nlp/models/llama/llama_attention.py (hunk 1 above)
Comment (tirthasheshpatel) [reply keras_nlp/models/llama/llama_attention.py]: Done.

File: keras_nlp/models/llama/llama_attention.py (hunk 4 above)
Comment (tirthasheshpatel) [reply keras_nlp/models/llama/llama_attention.py]: Done.

**Review by mattdangerw — APPROVED** [review]
LGTM! Will merge after testing finishes.
